File: web-platform/src/models/issue.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// ==================== Kanban Issue (card in the board) ====================

/// Issue card data as displayed in the kanban board.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct KanbanIssue {
    pub iid: u64,
    pub title: String,
    pub state: String,
    pub labels: Vec<String>,
    pub author: PlatformUser,
    pub assignees: Vec<PlatformUser>,
    pub created_at: String,
    pub updated_at: String,
    pub web_url: String,
    /// Number of associated MRs (only populated for in_progress column)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mr_count: Option<u64>,
}

// ==================== Issue Detail ====================

/// Full issue detail returned by the detail endpoint.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct IssueDetail {
    pub iid: u64,
    pub title: String,
    pub description: Option<String>,
    pub state: String,
    pub labels: Vec<String>,
    pub author: PlatformUser,
    pub assignees: Vec<PlatformUser>,
    pub milestone: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub closed_at: Option<String>,
    pub web_url: String,
    pub comment_count: u64,
    pub related_mrs: Vec<MergeRequestSummary>,
}

// ==================== Issue Summary (used in MR detail) ====================

/// Abbreviated issue info shown in MR detail's related_issues field.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct IssueSummary {
    pub iid: u64,
    pub title: String,
    pub state: String,
    pub web_url: String,
}

// ==================== MR Summary (used in Issue detail) ====================

/// Abbreviated MR info shown in issue detail's related_mrs field.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct MergeRequestSummary {
    pub iid: u64,
    pub title: String,
    pub state: String,
    pub author: PlatformUser,
    pub web_url: String,
}

// ==================== Platform User ====================

/// A user on the git platform (GitLab/GitHub).
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct PlatformUser {
    pub username: String,
    pub display_name: Option<String>,
    pub avatar_url: Option<String>,
}

File: web-frontend/src/types/issue.ts
// Phase 3 - Issue & MR types (snake_case per API spec, data from GitLab/GitHub)

export interface PlatformUser {
  username: string;
  display_name: string | null;
  avatar_url: string | null;
}

export interface Reviewer {
  user: PlatformUser;
  state: 'pending' | 'approved' | 'changes_requested';
}

// ==================== Issue Types ====================

export interface CreateIssueRequest {
  title: string;
  description?: string;
  labels?: string[];
  assignee?: string;
}

export interface AIGenerateRequest {
  prompt: string;
  title?: string;
  context?: string;
}

export interface IssueDetail {
  iid: number;
  title: string;
  description: string | null;
  state: 'opened' | 'closed';
  labels: string[];
  author: PlatformUser;
  assignees: PlatformUser[];
  milestone: string | null;
  created_at: string;
  updated_at: string;
  closed_at: string | null;
  web_url: string;
  comment_count: number;
  related_mrs: MergeRequestSummary[];
}

export interface MergeRequestSummary {
  iid: number;
  title: string;
  state: 'opened' | 'closed' | 'merged';
  author: PlatformUser;
  web_url: string;
}

export interface IssueSummary {
  iid: number;
  title: string;
  state: 'opened' | 'closed';
  web_url: string;
}

// ==================== Merge Request Types ====================

export interface MergeRequestDetail {
  iid: number;
  title: string;
  description: string | null;
  state: 'opened' | 'closed' | 'merged';
  author: PlatformUser;
  source_branch: string;
  target_branch: string;
  ci_status: 'pending' | 'running' | 'success' | 'failed' | 'canceled' | null;
  ci_web_url: string | null;
  review_status: 'pending' | 'approved' | 'changes_requested' | null;
  reviewers: Reviewer[];
  merge_status: 'can_be_merged' | 'cannot_be_merged' | 'checking' | 'unchecked';
  related_issues: IssueSummary[];
  additions: number;
  deletions: number;
  changed_files: number;
  created_at: string;
  updated_at: string;
  merged_at: string | null;
  web_url: string;
}

// ==================== SSE Event Types ====================

export interface SSEChunkEvent {
  type: 'chunk';
  content: string;
}

export interface SSEDoneEvent {
  type: 'done';
  content: string;
  title?: string;
}

export interface SSEErrorEvent {
  type: 'error';
  error: string;
  retCode?: string;
}

export type SSEEvent = SSEChunkEvent | SSEDoneEvent | SSEErrorEvent;
